src/Form/BasicOverview.php

```diff
@@ -0,0 +1,215 @@
+          '@time' => $date_formatter->format($user_tfa['saved']),
+          ':url'  => URL::fromRoute('tfa.disable', ['user' => $user->id()])->toString(),
+        ));
+      }
+      else {
+        $status_text = t('Status: <strong>TFA disabled</strong>, set @time.', array('@time' => $date_formatter->format($user_tfa['saved'])));
+      }
+      $output['status'] = array(
+        '#type'   => 'markup',
+        '#markup' => '<p>' . $status_text . '</p>',
+      );
+    }
+
+    if ($configuration['enabled']) {
+      // Validation plugin setup.
+      $enabled_plugin = $configuration['validate_plugin'];
+      $enabled_fallback_plugin = '';
+      if (isset($configuration['fallback_plugins'][$enabled_plugin])) {
+        $enabled_fallback_plugin = key($configuration['fallback_plugins'][$enabled_plugin]);
+      }
+
+      $output['app'] = $this->tfaPluginSetupFormOverview($enabled_plugin, $user, $user_tfa);
+
+      if ($enabled_fallback_plugin) {
+        // Fallback Setup.
+        $output['recovery'] = $this->tfaPluginSetupFormOverview($enabled_fallback_plugin, $user, $user_tfa);
+      }
+    }
+    else {
+      $output['disabled'] = [
+        '#type'   => 'markup',
+        '#markup' => '<b>Currently there are no enabled plugins.</b>',
+      ];
+    }
+
+    return $output;
+  }
+
+  /**
+   * Get TFA basic setup action links for use on overview page.
+   *
+   * @param string $plugin
+   *   The setup plugin.
+   * @param object $account
+   *   Current user account.
+   * @param array $user_tfa
+   *   Tfa data for current user.
+   *
+   * @return array
+   *   Render array
+   */
+  protected function tfaPluginSetupFormOverview($plugin, $account, array $user_tfa) {
+    $enabled = isset($user_tfa['status']) && $user_tfa['status'] ? TRUE : FALSE;
+
+    $output = array();
+    switch ($plugin) {
+      case 'tfa_totp':
+      case 'tfa_hotp':
+        $output = array(
+          'heading'     => array(
+            '#type'  => 'html_tag',
+            '#tag'   => 'h2',
+            '#value' => t('TFA application'),
+          ),
+          'description' => array(
+            '#type'  => 'html_tag',
+            '#tag'   => 'p',
+            '#value' => t('Generate verification codes from a mobile or desktop application.'),
+          ),
+          'link'        => array(
+            '#theme' => 'links',
+            '#links' => array(
+              'admin' => array(
+                'title' => !$enabled ? t('Set up application') : t('Reset application'),
+                'url'   => Url::fromRoute(
+                            'tfa.validation.setup',
+                            [
+                              'user'  => $account->id(),
+                              'method' => $plugin,
+                            ]
+                ),
+              ),
+            ),
+          ),
+        );
+        break;
+
+      case 'tfa_basic_sms':
```

Comment:
I will annonate this. That seems like the best way to make the whole method generic.
I already changed the route before to one single url where validation method is defined by the parameter `method`.